### DISCUSSION TITLE
Defining Actions and Previews

### DISCUSSION THREAD

**Discussion by jdhitsolutions** [mdgrs-mei/PowerShellRun#73] [General]
This may be two different topics, but they are related.  If I run `Invoke-PSRun` with minimal settings, I get the default behavior with a preview window and action keys.

![image](https://github.com/user-attachments/assets/8eb2bc1d-bf41-408a-a580-92ae73191293)

However, in my customized configuration, I am defining a new selector option so that I can define theme settings.

```powershell
$option = [PowerShellRun.SelectorOption]::new()
# <option changes>
Set-PSRunDefaultSelectorOption $option
```
However, this ends up turning off the preview. I found this setting and tried enabling it.

```powershell
$option.Theme.PreviewEnable = $True
```

But this doesn't seem to have any effect. I wonder if I need to define action keys, and if so how?  Part of my goal is to be able to create an additional preview action key for certain items. Can you provide some examples?

**Comment by mdgrs-mei** [comment]
```powershell
$option = [PowerShellRun.SelectorOption]::new()
$option.Theme.PreviewEnable = $false
Set-PSRunDefaultSelectorOption $option
```

This is the only way to disable the preview and calling `Set-PSRunDefaultSelectorOption` itself will not disable the preview window. So the behavior you described is not expected.

Probably is it related to updating the configuration in a session?

**Reply by jdhitsolutions** [reply]
In my configuration script I have code like this:

````
$option = [PowerShellRun.SelectorOption]::new()
$option.Theme.PreviewEnable = $True
$option.Theme.PromptForegroundColor = [PowerShellRun.FontColor]::BrightGreen

Set-PSRunDefaultSelectorOption -Option $option
```

The theme settings like color get applied when I run `Invoke-PSRun` but the `PreviewEnable` setting appears to be ignored.

**Comment by jdhitsolutions** [comment]
This is the full theming section of my configuration script.

```powershell

$roundBorder = [PowerShellRun.BorderSymbol]::new()
$roundBorder.TopLeft = '╔'  #'╭'
$roundBorder.TopRight = '╗' #'╮'
$roundBorder.BottomRight = '╝' #'╰'
$roundBorder.BottomLeft ='╚'   #'╯'
$roundBorder.Vertical = '║'
$roundBorder.Horizontal = '═'

$option = [PowerShellRun.SelectorOption]::new()
$option.Theme.PreviewEnable = $True
$option.Prompt = '🤓 > '
$option.Theme.Cursor = '▸ '
$option.Theme.CursorForegroundColor = [PowerShellRun.FontColor]::Cyan
$option.Theme.PreviewEnable = $false
$option.Theme.PreviewBorderForegroundColor = [PowerShellRun.FontColor]::BrightYellow
$option.Theme.CanvasBorderForegroundColor = [PowerShellRun.FontColor]::Yellow
$option.Theme.CanvasBorderSymbol = $roundBorder
$option.Theme.PreviewBorderSymbol = $roundBorder
$option.Theme.ActionWindowBorderSymbol = $roundBorder
$option.Theme.ActionWindowBorderForegroundColor = [PowerShellRun.FontColor]::BrightYellow
$option.Theme.SearchBarBorderForegroundColor = [PowerShellRun.FontColor]::BrightGreen
$option.Theme.PromptForegroundColor = [PowerShellRun.FontColor]::BrightGreen

Set-PSRunDefaultSelectorOption -Option $option
```

But something isn't getting set.

![image](https://github.com/user-attachments/assets/2247fa17-5198-41f4-9c0d-858064eca27f)

**Reply by mdgrs-mei** [reply]
Oh I see this line in your script after setting the CursorForegroundColor.

```powershell
$option.Theme.PreviewEnable = $false
```

**Comment by jdhitsolutions** [comment]
In a new session without my configuration script, I can see the preview is enabled.

![image](https://github.com/user-attachments/assets/ccdfb99a-c19e-4ed7-ba21-1b3db51e4fff)

Pasting my option commands in a new setting shows the desired setting.

![image](https://github.com/user-attachments/assets/64684ed1-ef2b-4d56-a9d9-8312f65e854c)

I wonder if this is a scoping issue.

**Comment by jdhitsolutions** [comment]
I see now that I was a little sloppy with my code. Preview is working now. So my next question is about customizing the action window.

![image](https://github.com/user-attachments/assets/fcc391e9-70a4-48b1-9183-3c524c960193)

I saw a reference in the README file about creating custom selectors, but it didn't give me enough information. For example, in the screen shot, is it possible to add an action to do something else with the selected item? In this case it is a folder, but I might want to customize an action for a file or other entry type.

**Reply by mdgrs-mei** [reply]
There is no way to customize the action window at the moment and I haven't came up with a good API yet to give the custom actions to the module.

`Invoke-PSRunSelectorCustom` allows you to customize what to show in the action window, but it's not the launcher, just a selector. You have to pass an array of entries and implement actions for the returned entry.

**Reply by jdhitsolutions** [reply]
Ok. This is a wish list item then.
